### PR TITLE
lodash: make object-based overloads higher than dictionary-based

### PR DESCRIPTION
**Pull request by Strate**
If we use `mapValues` with object, with all-equal types for all properties, it incorrectly infers that it is Dictionary

In this PR, i just move all object-based overloads higher than dictionary-based

### CODE REVIEW

File: types/lodash/lodash-tests.ts [hunk 1]
```diff
@@ -11886,6 +11886,24 @@ namespace TestMapKeys {
...
+
+        function mapNumber(x: number) {
+            return x
+        }
+
+        function expectValue(arg: Value) {
+            return arg
+        }
+
+        expectValue(_.mapValues({a: 1, b: 1, c: 1}, mapNumber))
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: A simpler (and more effective way) to test is this:

```ts
_.mapValues({a: 1, b: 1, c: 1}, (value) => 2); // $ExpectType { a: number; b: number; c: number; }
```

**Review by aj-r — CHANGES_REQUESTED** [review]
The problem is, moving the object overload ahead makes the dictionary overloads effectively useless. All dictionaries will use the object overload instead, so you might as well delete the dictionary overloads.

Thoughts on the test failures:

- Some of the test failures look OK - like getting `{ [x: string]: string; }` instead of `Dictionary<string>` is fine, since they are effectively the same type. You just need to change the `$ExpectType` to use the new type.
- The "error" on line 11743 actually looks preferable. Again, you just need to update the `$ExpectType` on that line.
- The remaining tests (`expected Dictionary<string>, got {}`) look concerning. You should probably find a way to fix these tests before merging.

If you do find a way to make this work nicely for both object literals and dictionaries, then don't forget to update the wrapper versions of `mapValues`, too.

**Comment by Strate** [comment]
@aj-r thank you for your response!
The most surprising for me is failing tests of lowdb - I don realized yet reason of failure. Other cases will be fixed of course.

**Comment by Strate** [comment]
The problem is about numeric dictionary: typescript cannot play well with mapped types for numeric dictionary: https://github.com/Microsoft/TypeScript/issues/22105#issuecomment-375208290

Seems that it is impossible to resolve before ts fix

**Comment by aj-r** [comment]
Don't worry about the lowdb failure. It's caused by an unrelated issue in typescript@next.

**Comment by aj-r** [comment]
@Strate sorry, we just merged a PR that conflicts with your changes. You might want to pull the latest code from DefinitelyTyped/master.

On the bright side, I've found something that seems to work:

- 1st overload: `NumericDictionary` only
- 2nd overload: `T extends object`

Note that we have no `Dictionary` overload because it will be matched by the `NumericDictionary` overload. This means that the `NumericDictionary<T>` overloads must unfortunately return `Dictionary<T>` (same as they did before).

The important thing here is that object literals will NOT be matched by the `NumericDictionary` overload. So basically here's what we need to do:

- Keep the overloads in the same order as they were before (so probably revert your change)
- Replace `Dictionary<T> | NumericDictionary<T>` with just `NumericDictionary<T>` in all overloads of `mapValues`.

**Comment by Strate** [comment]
@aj-r big thanks on this, I'll try your solution soon.

File: types/lodash/lodash-tests.ts [hunk 2]
```diff
@@ -5651,6 +5651,14 @@ fp.now(); // $ExpectType number
     _.mapValues(abcObject); // $ExpectType AbcObject
     _.mapValues(abcObjectOrNull); // $ExpectType Partial<AbcObject>
 
+    // $ExpectType { a: number; b: number; c: number; }
+    _.mapValues({ a: 1, b: 1, c: 1 }, (value, key, collection) => {
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: The test failure is really weird - it looks like it's not getting the right types for `value` and `collection`. If I modify this code to explicitly pass the type parameters, it works file (e.g. `_.mapValues<{ a: number, b: number, c: number }, number>({ a: 1, b: 1, c: 1 }, (value, key, collection) => {`).

Also, explicitly setting the callback parameter types seems to work, so maybe you should do something like this:

```ts
// $ExpectType { a: string; b: string; c: string; }
_.mapValues({ a: 1, b: 1, c: 1 }, (value: number, key: string, collection: { a: number, b: number, c: number }) => "");
```

**Review by aj-r — CHANGES_REQUESTED** [review]
Don't forget to update the wrapper functions (`LoDashImplicitWrapper` and `LoDashExplicitWrapper`), and add wrapper versions of your new test.

Also make sure you update the `fp` functions. To do this,  see `readme.md`. Here's the tl;dr version:
- Open a terminal in `types/lodash/scripts`
- `npm install`
- `npm run fp`

File: types/lodash/lodash-tests.ts (hunk 2 above)
Comment (aj-r) [reply types/lodash/lodash-tests.ts]: Looks like the test failure is caused by a typescript issue i logged a while ago: TypeScript#21525. Looks like the typescript team has no intention of fixing this.

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.

**Pull request closed without merging**